File: shared/domain/onboarding.ts
export type AgreementKey = 'terms' | 'privacy' | 'marketing';

export interface AgreementState {
  terms: boolean;
  privacy: boolean;
  marketing: boolean;
}

export interface ProfileDraft {
  name: string;
  birthDate: string;
  gender: 'M' | 'F' | 'N' | null;
  heightCm: string;
  weightKg: string;
  province: string | null;
  district: string | null;
}

export function setAllAgreements(checked: boolean): AgreementState {
  return {
    terms: checked,
    privacy: checked,
    marketing: checked,
  };
}

export function toggleAgreement(
  state: AgreementState,
  key: AgreementKey,
): AgreementState {
  const next = { ...state, [key]: !state[key] };
  return next;
}

export function areRequiredAgreementsAccepted(state: AgreementState): boolean {
  return state.terms && state.privacy;
}

export function areAllAgreementsAccepted(state: AgreementState): boolean {
  return state.terms && state.privacy && state.marketing;
}

export function normalizeBirthYear(value: string): number | undefined {
  const trimmed = value.trim();
  const yearOnlyMatch = trimmed.match(/^(\d{4})$/);
  const compactDateMatch = trimmed.match(/^(\d{4})(\d{2})(\d{2})$/);
  const separatedDateMatch = trimmed.match(/^(\d{4})([-/.])(\d{1,2})\2(\d{1,2})$/);

  if (yearOnlyMatch) {
    const year = Number(yearOnlyMatch[1]);
    return isReasonableBirthYear(year) ? year : undefined;
  }

  const dateMatch = compactDateMatch ?? separatedDateMatch;
  if (!dateMatch) return undefined;

  const year = Number(dateMatch[1]);
  const month = Number(dateMatch[compactDateMatch ? 2 : 3]);
  const day = Number(dateMatch[compactDateMatch ? 3 : 4]);
  if (!isReasonableBirthYear(year)) return undefined;
  if (!isValidPastOrPresentDate(year, month, day)) return undefined;

  return year;
}

function isReasonableBirthYear(year: number): boolean {
  const currentYear = new Date().getFullYear();
  return Number.isInteger(year) && year >= 1900 && year <= currentYear;
}

function isValidPastOrPresentDate(year: number, month: number, day: number): boolean {
  if (!Number.isInteger(month) || !Number.isInteger(day)) return false;
  const date = new Date(year, month - 1, day);
  if (
    date.getFullYear() !== year ||
    date.getMonth() !== month - 1 ||
    date.getDate() !== day
  ) {
    return false;
  }

  const today = new Date();
  const todayStart = new Date(today.getFullYear(), today.getMonth(), today.getDate());
  return date <= todayStart;
}

export function canContinueProfile(draft: ProfileDraft): boolean {
  return Boolean(
    draft.name.trim() &&
    normalizeBirthYear(draft.birthDate) &&
    draft.gender &&
    draft.province &&
    draft.district
  );
}

File: apps/android-native/app/src/main/java/com/wxxtae/pws/nativepreview/domain/Onboarding.kt
package com.wxxtae.pws.nativepreview.domain

import java.time.LocalDate

enum class AgreementKey {
    Terms,
    Privacy,
    Marketing,
}

data class AgreementState(
    val terms: Boolean = false,
    val privacy: Boolean = false,
    val marketing: Boolean = false,
)

enum class Gender {
    Male,
    Female,
    None,
}

data class ProfileDraft(
    val name: String = "",
    val birthDate: String = "",
    val gender: Gender? = null,
    val heightCm: String = "",
    val weightKg: String = "",
    val province: String? = null,
    val district: String? = null,
)

object OnboardingRules {
    private val yearOnlyPattern = Regex("^(\\d{4})$")
    private val compactDatePattern = Regex("^(\\d{4})(\\d{2})(\\d{2})$")
    private val separatedDatePattern = Regex("^(\\d{4})([-/.])(\\d{1,2})\\2(\\d{1,2})$")

    fun setAllAgreements(checked: Boolean): AgreementState = AgreementState(
        terms = checked,
        privacy = checked,
        marketing = checked,
    )

    fun toggleAgreement(state: AgreementState, key: AgreementKey): AgreementState = when (key) {
        AgreementKey.Terms -> state.copy(terms = !state.terms)
        AgreementKey.Privacy -> state.copy(privacy = !state.privacy)
        AgreementKey.Marketing -> state.copy(marketing = !state.marketing)
    }

    fun areRequiredAgreementsAccepted(state: AgreementState): Boolean = state.terms && state.privacy

    fun areAllAgreementsAccepted(state: AgreementState): Boolean = state.terms && state.privacy && state.marketing

    fun normalizeBirthYear(value: String, today: LocalDate = LocalDate.now()): Int? {
        val trimmed = value.trim()
        yearOnlyPattern.matchEntire(trimmed)?.let { match ->
            val year = match.groupValues[1].toInt()
            return year.takeIf { isReasonableBirthYear(it, today) }
        }

        compactDatePattern.matchEntire(trimmed)?.let { match ->
            return normalizeBirthDate(
                year = match.groupValues[1].toInt(),
                month = match.groupValues[2].toInt(),
                day = match.groupValues[3].toInt(),
                today = today,
            )
        }

        separatedDatePattern.matchEntire(trimmed)?.let { match ->
            return normalizeBirthDate(
                year = match.groupValues[1].toInt(),
                month = match.groupValues[3].toInt(),
                day = match.groupValues[4].toInt(),
                today = today,
            )
        }

        return null
    }

    fun canContinueProfile(draft: ProfileDraft, today: LocalDate = LocalDate.now()): Boolean =
        draft.name.trim().isNotEmpty() &&
            normalizeBirthYear(draft.birthDate, today) != null &&
            draft.gender != null &&
            !draft.province.isNullOrBlank() &&
            !draft.district.isNullOrBlank()

    private fun normalizeBirthDate(year: Int, month: Int, day: Int, today: LocalDate): Int? {
        if (!isReasonableBirthYear(year, today)) return null
        val date = runCatching { LocalDate.of(year, month, day) }.getOrNull() ?: return null
        if (date.isAfter(today)) return null
        return year
    }

    private fun isReasonableBirthYear(year: Int, today: LocalDate): Boolean = year in 1900..today.year
}
